[MU3] reduce useless whitespace in JSON files; better way to update libmscore/sym.* from SMuFL (including bugfixes)

mirabilos opened the pull request:
Split-off PR, this one now only deals with the JSON and SMuFL topics; related MU4 PR is https://github.com/musescore/MuseScore/pull/4883

----

These are the first fruits from [my attempt to replace the Emmentaler accidentals with those from Parnassus in MScore](https://github.com/mirabilos/MuseScore/tree/parnassus-accidentals). These changes _do **not**_ yet change the font itself.

First off, I found over 12½ KB of useless whitespace at end-of-line in the font and SMuFL JSON files which getting rid of reduces the whole application size (by about 37K because Qt interns both latin1 and Unicode forms).

[…]

We could save even more space in the JSONs by re-indenting them to e.g. two spaces (as ECMA-262 suggests) or even one tab per level. Do tell me if you wish for me to update the PR to do just that.

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [ ] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

Jojo-Schmitz commented:
We do take the SMuFL directly and as is from https://github.com/w3c/smufl, so better get these changes in there, else we''ll have merge conflicts when updating, like with #6543 (merged into 3.x) and #4883 (pending merge into master).

Jojo-Schmitz commented:
Hmm, maybe for 3.x this SMuFL change is OK, considering that SMuFL 1.4 most probably will never make it into 3.x, as 3.6 will be the last of that series. This change should not make it into master though until it comes from SMuFL directly.

mirabilos commented:
@Jojo-Schmitz good point, but the whitespace change (even the more daring one I suggested) can be automated; I’m assuming we don’t change those files ourselves.

I’d suggest doing so at build time, but considering we also build on Windows, that’s going to be tricky since it doesn’t even have `sed`.

I guess I’ll leave this in the 3.x PR as-is for now, and will just omit that part from the master PR then.

Jojo-Schmitz commented:
How much space would the _more daring one_ buy us? If it is 37k for those single (?) trailing spaces, it got to be 2 times more for a 2 space indent and 4 times more for a tab indent, right? So up to 5*37k, quite a lot!

Jojo-Schmitz commented:
And if we manage to do that at build time, it'd be even more, as we could eliminate all indentation.

mirabilos commented:
Yes, it’s a good 19.7% in total. I filed a PR against SMuFL with this: https://github.com/w3c/smufl/pull/135

This would save about 346K in the mscore binary. Granted, not *that* much compared to the total size, but the parser overhead would be even less.

(We could even strip the indent fully for the embedded version, saving a further 3*34½ kB.)

Jojo-Schmitz commented:
As https://github.com/w3c/smufl/pull/135 git merged into SMuFL, please adjust your's here to match that one ;-)

mirabilos commented:
Adjusted accordingly @Jojo-Schmitz ☻

Jojo-Schmitz commented:
Does this include the changes of all 3 commits of that SMuFL PR?

Jojo-Schmitz commented:
And please don't do this for master, I'll do it as part of #4883, and take it from SMuFL directly

mirabilos commented:
@Jojo-Schmitz it includes all these changes, but not (yet) the “follwed” fix. Should I add that, too?

Yes, I will not do that for master, only submit the code correction there.

Jojo-Schmitz commented:
Yes, the full package of the SMuFL PR please

mirabilos commented:
OK, updated with the SMuFL spelling fix, so these are now bitwise identical to the latest SMuFL

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -3473,14 +3470,15 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       QT_TRANSLATE_NOOP("symUserNames", "Dashed barline"),
       QT_TRANSLATE_NOOP("symUserNames", "Dotted barline"),
       QT_TRANSLATE_NOOP("symUserNames", "Double barline"),
+      "Final barline",
```
Yes, please these:
```
    QT_TRANSLATE_NOOP("symUserNames", "Final barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Heavy barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Heavy double barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Reverse final barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Short barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Single barline"),
    QT_TRANSLATE_NOOP("symUserNames", "Tick barline"),
```

Jojo-Schmitz replied:
I'll add those missing `QT_TRANSLATE_NOOP()`s in another PR, dealing with those additional barlines

mirabilos replied:
We’ll need to add them to the JSON file as well, so better do it here and let #6752 depend on this one.

Jojo-Schmitz commented on fonttools/smufl2sym-in-trans.jsn:
```diff
@@ -0,0 +1,341 @@
+{
+  "accSagittal11LargeDiesisDown": null,
```
What is the purpose if this file? Why .jsn and not .json?

mirabilos replied:
As I documented in the commit message, this file contains all keys for which a QT_TRANSLATE_NOOP is to be generated.

mirabilos replied:
And .jsn is because I grew up on MS-DOS with 8.3 filenames and prefer them… but if this is a sticking point, I can rename it.

Jojo-Schmitz replied:
> As I documented in the commit message, this file contains all keys for which a QT_TRANSLATE_NOOP is to be generated.

Ah, sorry missed this reply... something to be very careful about, we'd need to remember adding those here **and** modifying in sym.cpp, or add just here **and** run the script, rather than adding **just** modifing in sym.cpp (as later run the script would revert that change).
Probably worth a comment in sym.cpp (and maybe even that comment added via this script, along the lines of "don't update the strings here manually, these are auto generated by ..."

Jojo-Schmitz commented:
Wouldn't the `metadata.json` files (for at least 'our' fonts, for Bravura and Petaluma it might be worth a PR against those, maybe?) also benefit from the whitespace cleanup?

Jojo-Schmitz commented:
We should rather not touch `Petaluma/Metadat.json` and `Bravura/metada.json` though, not without Steinberg's buy-in

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -2842,6 +2844,8 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symNames = { {
       "windWeakAirPressure",
+      // SMuFL standard symbol IDs }}}
+
 
```
one empty line too much for my taste

mirabilos replied:
The empty line makes it easier to see where the SMuFL-derived stuff begins and ends. (Yes, we also have the comment, but…)

If you insist, I can remove it.

mirabilos replied:
Oh wait, I see. Sorry. Fixed.

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -3173,19 +3173,17 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
-      QT_TRANSLATE_NOOP("symUserNames", "Combining raise by one 17-limit schisma)"),
-      QT_TRANSLATE_NOOP("symUserNames", "Combining raise by one 19-limit schisma)"),
+      QT_TRANSLATE_NOOP("symUserNames", "Combining raise by one 17-limit schisma"),
+      QT_TRANSLATE_NOOP("symUserNames", "Combining raise by one 19-limit schisma"),
```
These two are my brown bag bugs :-/

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -4157,8 +4155,8 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
-      "Toe-dtep",
-      "Toe-dap",
+      "Toe-step",
+      "Toe-tap",
```
These two are likely my brown bag bugs :-/

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -4947,9 +4945,8 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       "Oblique straight line SW-NE",
       "Oblique straight line tilted NW-SE",
       "Oblique straight line tilted SW-NE",
-      QT_TRANSLATE_NOOP("symUserNames", "Short trill"),
```
This is my brown bag bug :-/

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -4985,12 +4982,13 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       "Shake (Muffat)",
       "Short oblique straight line NW-SE",
       "Short oblique straight line SW-NE",
+      QT_TRANSLATE_NOOP("symUserNames", "Short trill"),
```
This is my brown bag bug :-/

Jojo-Schmitz commented:
See also https://gist.github.com/lasconic/be08234b3dd074e99edc1fb060f0abcb for a Python script doing those conversions. Not as  complete as your script, but maybe more portable?

mirabilos commented:
The Python script, unfortunately, not only barely does anything, it also doesn’t do that right (e.g. the Unicode characters, as you saw). It also doesn’t do the QT_TRANSLATE_NOOP thing. jq is pretty portable, and other than that it just needs sh (even the one from the Weihenstephan tools should suffice).

I even wrote the things in a way to depend on _only_ jq, not use sed and other tools, so we can even add a .bat file doing the same thing with jq for Windows…

mirabilos commented:
@Jojo-Schmitz I updated the PR to:

- not touch Pataluma (I already had kept fingers off Bravura)
- add those barlines as QT_TRANSLATE_NOOPs

Yes, the font JSON files would also benefit from the whitespace changes I submitted to SMuFL. But I think the other approach you suggested (to drop all optional whitespace at build time) is the generally better one.

Jojo-Schmitz commented:
Thanks
Considering how fast your SMuFL PR got accepted, it might be worthwhile to send a PR to Steinberg for Petaluma and Bravura too

mirabilos commented:
Joachim Schmitz dixit:

>Considering how fast your SMuFL PR got accepted, it might be worthwhile
>to send a PR to Steinberg for Petaluma and Bravura too

OK will do

mirabilos commented:
Updated to
- rename the `.jsn` file to `.json`
- split the [jq](https://stedolan.github.io/jq/) scripts into `.jq` files
  - make `smufl2sym.sh` just call those (GNU/Linux, BSD, Mac OSX)
  - add `smufl2sym.bat` doing the same on Windows, as requested by @Jojo-Schmitz 
    - depends on jq (see above) and [dos2unix](https://waterlan.home.xs4all.nl/dos2unix.html)
    - dependencies are documented in a comment at the top of the batch script
- removed the dependency on column(1)

Output is bytewise identical on GNU/Linux and Windows.

mirabilos commented:
Bravura PR: https://github.com/steinbergmedia/bravura/pull/36

Petaluma PR: https://github.com/steinbergmedia/petaluma/pull/6

mirabilos commented:
Updated with the new QT_TRANSLATE_NOOPS from commit b69e2a75d3639ea7aa9b31f1d9d0281d630fbf8b and commit 1f5ec3fc1d920bc4617efd9b81ff6df38051341f (no other change; rebasing on top of latest 3.x shows the resulting sym.cpp identical modulo a whitespace fix)

mirabilos commented:
Joachim Schmitz dixit:

>Ah, sorry missed this reply... something to be very careful about, we'd
>need to remember adding those here **and** modifying in sym.cpp, or add
>just here **and** run the script, rather than adding **just** modifing
>in sym.cpp (as later run the script would revert that change).

I could move the generated parts out of sym.{cpp,h} and into separate
files that indicate they are generated and then #include these.

We could conceivably generate them at every build, but that’s overkill
plus requires jq, so, better to check them in I guess.

>Probably worth a comment in sym.cpp (and maybe even that comment added
>via this script, along the lines of "don't update the strings here
>manually, these are auto generated by ..."

Like that what I wrote above?

Jojo-Schmitz commented:
I'd leave sym.{cpp,h} as they are, just add that one line comment to the start of those, next to that `SMuFL standard symbol IDs`, making it `SMuFL standard symbol IDs, generated automagigally, so don't edit manually but use the script at ... instead`

mirabilos commented:
Joachim Schmitz dixit:

>I'd leave sym.{cpp,h} as they are, just add that one line comment to
>the start of those, next to that `SMuFL standard symbol IDs`, making it
>`SMuFL standard symbol IDs, generated automagigally, so don't edit
>manually but use the script at ... instead`

Hm, okay. I guess you can just put that comment in the three places
in these files manually, if it’s above the “SMuFL standard symbol IDs”
it’ll be outside of what the script generates and thus kept.

mirabilos commented:
@Jojo-Schmitz https://github.com/musescore/MuseScore/compare/ff733f2b050d70f650c816746c40d81c79a7a76f..c27d0317b8f9970d7a8f353f2bdace1a9c383bed what do you think?

Can you do the same for the MU4 PR #4883 ?

Jojo-Schmitz commented:
Yes. Tomorrow...

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -51,6 +51,12 @@ std::array<uint, size_t(SymId::lastSym)+1> ScoreFont::_mainSymCodeTable { {0} };
 const std::array<const char*, int(SymId::lastSym)+1> Sym::symNames = { {
       "noSym",
+
+      // DO NOT edit the SMuFL standard symbol IDs (see below) manually!
```
`symbols IDs` vs. `symbol names`?

Jojo-Schmitz commented on libmscore/sym.cpp:
```diff
@@ -2879,14 +2886,19 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symNames = { {
 const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       QT_TRANSLATE_NOOP("symUserNames", "No symbol"),
+
+      // DO NOT edit the SMuFL standard symbol IDs (see below) manually!
```
`symbols IDs` vs. `symbol user names`?

mirabilos commented:
Joachim Schmitz dixit:

>+      // DO NOT edit the SMuFL standard symbol IDs (see below) manually!
>
>`symbols IDs` vs. `symbol names`?

This was what the original had… okay, the original had them
only in sym.h… maybe “the order of these items is from the
SMuFL IDs” is a justification? ;-)

Jojo-Schmitz commented:
Check https://github.com/musescore/MuseScore/pull/4883/commits/8d56d717ed84cb1582cad856c3166391fe5b7471 and https://github.com/musescore/MuseScore/pull/4883/commits/b736fc0decf690c083d5d61ef092598f806680d9

mirabilos commented:
Updated this one accordingly (modulo the “namess” typo).

Jojo-Schmitz commented:
Typo fixed, check https://github.com/musescore/MuseScore/pull/4883/commits/7021bc9d9f38c49e45bf0ac2461019644ddd7382
